hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordMerger.java

```diff
@@ -63,8 +63,15 @@ public interface HoodieRecordMerger extends Serializable {
    * It'd be associative operation: f(a, f(b, c)) = f(f(a, b), c) (which we can translate as having 3 versions A, B, C
    * of the single record, both orders of operations applications have to yield the same result)
    * This method takes only full records for merging.
+   *
+   * @param older     Older record in terms of commit time ordering.
+   * @param oldSchema The schema of the older record.
+   * @param newer     Newer record in terms of commit time ordering.
+   * @param newSchema The schema of the newer record.
+   * @param props     The additional properties for the merging operation.
+   * @return The merged record and schema. The record is expected to be non-null. If the record represents a deletion, the operation must be set as {@link HoodieOperation#DELETE}.
    */
-  Option<Pair<HoodieRecord, Schema>> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException;
+  Pair<HoodieRecord, Schema> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException;
```

Comment:
if we just takes two buffered records, things would be much easier:

```java
BufferedRecord merge(BufferedRecord older, BufferedRecord newer, RecordContext context);
```

the benefits to do this:

- BufferedRecord is the core abstraction for merging, it is more easier to evolve in the future;
- avoid unnecessary hoodie record -> buffered record -> hoodie record conversion;
- more clear impl/semantics for deletes handling in mergers.

We just need to rename the `BufferedRecord` to be a more general name when it been used as a user API.

Reply:
On the same topic of, whether we need a new Pojo here or  is `HoodieRecord` the right choice. HoodieRecord is used here as a change record i.e we merge multiple change records in commit time order, based on the custom logic.. Is this use of the pojo consistent with other places ..

Reply:
There may be some practical limitations that will prevent this from happening though. For example, there is no way to currently pass in the payload information to the `HoodieAvroRecordMerger` so we would not know how to construct the payloads properly.

Reply:
> there is no way to currently pass in the payload information to the HoodieAvroRecordMerger

We can, just let the T in BufferedRecord be the payload instance?

Reply:
Discussed offline and for payload based merging, the T should be avro `IndexdRecord`, either think though a way to invoke the deprecated payload API or just ignore it and impl event time based merging on it.